A prototype of first-class functions [DON'T MERGE]

pearu opened the pull request:
Tackles the issue #3405 but closes once support for `njit` decorated functions is implemented.

The middle-level description of the first-class function model is as follows:
1. A function, as an instance of **the first-class type**, is represented by its *address*, *signature*, and a pointer to a *parent* python object. The address enables accessing the compiled implementations of the function as well as to enable using function instances as items within njitted functions. The signature enables constructing the LLVM IR level call to the compiled functions. The parent object is used as a reference to the high-level implementation (python function, for instance) that enables compiling the function for a different set of argument types in just-in-time manner as well as it is used as return value of the first-class function.
2. **Wrapper address protocol**: When a first-class function is passed into a njitted function, it needs to be "unboxed". The unboxing requires a method for acquiring the address of the compiled function. For instance, if the input argument is a `numba.cfunc(<signature>)` decorated function (of type `CFunc`), then the address is readily available via its `_wrapper_address` attribute. This PR introduces so-called *wrapper address protocol* (see https://github.com/numba/numba/pull/4967/files#diff-275bdd7060d042bc811293de00e7878dR115-R143 for the definition and support) that enables interpreting arbitrary python objects as first-class functions when passed to a njitted function as an argument. For instance, to retrieve a function address from a python object such as `CFunc`, the function `numba.function._get_wrapper_address(obj, signature)` is called from the unboxing procedure. The implementation of the `_get_wrapper_address` for the `CFunc` objects would be as simple as (assuming that `CFunc` instance is compiled against a correct signature):
```
def _get_wrapper_address(obj, signature):
    return obj._wrapper_address
```
3. **Extensibility**: The wrapper address protocol model allows for extending first-class function model to arbitrary objects (not just for `cfunc` decorated functions). For instance, one can implement caching of compiled functions as an independent method to numba jit caching model, or one can call arbitrary library functions from njitted functions (as long as one is able to determine the pointer values of the library functions, one can use `ctypes` for that). See https://github.com/numba/numba/pull/4967/files#diff-75b50def42de3256e721ce2c6e79c47dR282-R322 for an example of how to call a libc function `time` from a njitted function.

Current status:

- [next PR] Pass functions as arguments to njitted functions:
  - [x] `cfunc` decorated functions
  - [x] `njit` decorated functions
  - [next PR] pure Python functions
  - [next PR] ctypes functions
- [next PR] Use functions as items within njitted functions:
  Both constant (available via namespace lookup) and passed in functions (available as arguments of njitted functions)
  - [x] `cfunc` decorated functions
  - [x] `njit` decorated functions
  - [next PR] pure Python functions
  - [next PR] ctypes functions
- [next PR] Use functions as return values from njitted functions
  Both constant and passed in functions
  - [x] `cfunc` decorated functions
  - [x] `njit` decorated functions
  - [next PR] pure Python functions
  - [next PR] ctypes functions
- [next PR] Overload function calls within njitted functions
  Both constant and passed in functions
  - [next PR] `cfunc` decorated functions
  - [x] `njit` decorated functions
- [next PR] Constuct function signatures
  - [x] int64 arguments
  - [x] other argument types
  - [next PR] tests, partially blocked by `Signature` vs `FunctionType` issue
- [x] Unit-tests
- [next PR] Documentaton

Example:
```
from numba import cfunc, njit, int64

@cfunc(int64(int64))
def a(i):
    return i + 1

@cfunc(int64(int64))
def b(i):
    return i + 2

@njit
def foo(f, g):
    i = f(2)
    seq = (f, g)
    for fun in seq:
        i += fun(i)
    return i

a_ = a._pyfunc
b_ = b._pyfunc
assert foo(a, b) == a_(2) + a_(a_(2)) + b_(a_(2) + a_(a_(2)))
```

sklam commented on numba/typing/templates.py:
```diff
@@ -113,6 +113,13 @@ def as_function(self):
         sig = signature(self.return_type, *((self.recvr,) + self.args))
         return sig
 
+    def __call__(self, *args):
```
Why is this necessary? Signature can't really numba type.

pearu replied:
This is a convenience method for constructing signatures that involve function types as return values. (When trying it now, I see it is broken but) as an example, consider
```
from numpy import *
int32(int32)   # this results a signature `(int64,) -> int64`
```
With the first-class function types, we also need to construct a signature where the return type is a function type. For instance, it would be convenient to construct the corresponding signature as follows 
```
rtype = int32(int32)
sig = rtype(int32)           # here the __call__ methods is effective
```
that would be a signature of a function taking int32 value as input and returning a function of type `int32(int32)`.

It is in some sense unfortunate that `int32(int32)` returns a Signature object. With the first-class function support, it would be more natural if it would return numba `FunctionType` instance.

sklam replied:
> it would be more natural if it would return numba FunctionType instance.

Agree, we need to look at that as part of our Q1 housekeeping.

pearu replied:
Removed `__call__` method for now.
TODO for Q1 or later: `Type.__call__` returns `TypeFunction` instance. The question of how this can be transformed to `Signature` instance, needs an answer. E.g. is `Signature` still required as `TypeFunction` carries the same info?

stuartarchibald commented:
/AzurePipelines run

pearu commented:
The current PR implements first-class function type support from `cfunc` decorated functions.
Some features work also for `njit` decorated functions but the full support will be implemented in a separate PR.
@sklam , please review.
The

pearu commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,585 @@
+
+        self.assertEqual(w(njit, a, b), w(lambda f:f, a.pyfunc, b.pyfunc))
+
+    def test_all(self):
```
The TestFunctionType class currently implements 17 usage patterns (methods starting with `_test`) of first-class functions that could be of 6 different kinds (see `all_func_kinds` list).
The `test_all` method applies the test methods to all different input types so that there will be 102 tests run in total. Although this PR implements full support for one kind only (`cfunc_func`), it seemed impractical and error-prone to have the same test pattern repeated possibly 6 times.

Ideally, the above should be implemented using parameterized test methods (e.g. [pytest.mark.parametrize](https://docs.pytest.org/en/latest/parametrize.html#parametrize)) but std unittest does not support this. There is [parameterized](https://github.com/wolever/parameterized) project for unittest though.

Ahh, I just found `https://docs.python.org/3/library/unittest.html#subtests`... and fix the above.

pearu replied:
Fixed. The tests are now explicit in what function types are supported and what not-yet supported.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,210 @@
+from numba.ccallback import CFunc
+from numba import cgutils
+from llvmlite import ir
+# from numba.dispatcher import Dispatcher
```
Fixed.

pearu commented on numba/typing/context.py:
```diff
@@ -175,6 +175,9 @@ def explain_function_type(self, func):
                 defns.extend(getattr(tpl, 'cases', []))
 
+        elif isinstance(func, types.FunctionType):
+            defns.append(func)
```
Removed.

pearu commented on numba/types/function.py:
```diff
@@ -0,0 +1,296 @@
+            # TODO: analyze the connection to the unboxing model,
+            # why does it just works? See also get_call_type method below.
+            import numba
+            self.ftype = numba.njit(ftype)
```
Correct. This codelet will be likely relevant for other first-class function kinds but not here. Will remove.

pearu commented on numba/targets/base.py:
```diff
@@ -525,7 +525,7 @@ def get_function(self, fn, sig, _firstcall=True):
         sig = sig.as_function()
         if isinstance(fn, (types.Function, types.BoundFunction,
-                           types.Dispatcher)):
+                           types.Dispatcher, types.FunctionType)):
```
Fixed.

pearu commented on numba/types/function.py:
```diff
@@ -0,0 +1,296 @@
+    """
+
+
+def numbatype(obj):
```
The important feature of `numbatype` is that it supports the string representation of function types that contain other function types as arguments or return values. Consider, for instance,
```
numbatype('int32(int32) (float64(int64))')
```
that is a function type representing a function takes a function as input and returns another function. If `numbatype` would only handle `Signature` and `Type` arguments then the following would work:
```
>>> numbatype(int32(int32))(numbatype(float64(int64)))
(float64(int64),) -> int32(int32)
```
but for users, constructing such signature would be easier when converting from a string as shown in the first example.

Ideally, one would use
```
int32(int32) (float64(int64))
```
but this requires either making `Signature` a callable or unifying `Signature` and `FunctionType`.

Re string to `Type` conversion: the RBC project implements the string specification of types in full generality if the five basic types are void, atomic, pointer, struct, and function.
See https://github.com/xnd-project/rbc/blob/master/rbc/typesystem.py#L235-L295 .

In sum, this is a user-convenience question and could be addressed later. Thoughts?

pearu replied:
See the `sig` comment below as it turns out that supporting strings avoids the need to lower `Type` instances.

pearu replied:
I think I can eliminate `numbatype` altogether.

pearu replied:
`numbatype` is eliminated. Introduced `Signature.as_type` method that returns function type.

pearu commented on numba/types/function.py:
```diff
@@ -0,0 +1,296 @@
+
+        Parameters
+        ----------
+        sig : str
```
The call to `__wrapper_address__` is constructed from `unbox_function_type` function and to support non-string `sig` would mean that the `Type` instances would be needed to lowered as well.
Using a string to represent a `Type` value at the LLVM IR level is much easier and sufficient (as long as this is needed only for passing `Type` values, not for manipulating the values within LLVM IR level).

pearu replied:
I wonder if the approach using `pyapi.serialize_object` could avoid the need to lower `Type` instances?
It would be equivalent to `sig` being string except that the `__wrapper_address__` caller unserializes the pickled data to `FunctionType` and passing it to `__wrapper_address__`.

pearu replied:
I confirm, using `pyapi.serialize_object` allows `sig` to be `Signature`.

pearu replied:
Fixed, the signature must now be `Signature` instance.

pearu commented:
@sklam @stuartarchibald I have addressed all Siu's notes:
1. replace test_all method with explicit test methods.
2. eliminate `numbatype` function as not needed here.
3. signatures are expected to be given as `Signature` instances 
The PR is ready for review.

stuartarchibald commented:
Thanks @pearu I've marked it as such. I'll hopefully be able to look at it in the next day or so.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+        f.pyfunc = func
+        return f
+    raise NotImplementedError(
+        f'ctypes decorator for {func} with signature {sig}')
```
I think that all these should have `sig=None` as default. This is because some of them don't even use the signature and it should be asserted that the `signature` is indeed `None` in those cases, this to prevent mistakes in future use.

pearu replied:
The `sig` argument was introduced in order to have a family of decorators available for testing.
Currently, we only test functions with signature `int64(int64)` but for njit decorated first-class functions support other signatures need to be tested as well.
To resolve this, I suggest making the usage of `sig` argument explicit and eliminate it where possible.

pearu replied:
Fixed.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+                  njit2_func, ctypes_func, wap_func]
+
+
+class TestFuncionType(TestCase):
```
```suggestion
class TestFunctionType(TestCase):
```

pearu replied:
Fixed.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+
+
+def wap_func(func, sig=int64(int64)):
+    return WAP(func, sig)
```
Same comment about `sig=None`.

pearu replied:
Fixed.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+        return self.sig
+
+    def __call__(self, *args, **kwargs):
+        return self.pyfunc(*args, **kwargs)
```
Should this be available as a mixin for public use?

pearu replied:
Perhaps. `WAP` provides a way to turn any Python function to a first-class function as long as decorating it with `cfunc` makes sense.

pearu replied:
Actually, this issue is about supporting pure Python functions as first-class function type objects and is a little out-of-scope of this PR that introduces the feature mostly in terms of `cfunc` compiled functions. I'd like to postpone discussing this feature to the next PR as I think there is a cleaner way to support pure Python functions.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+        a_ = a._pyfunc
+        b_ = b._pyfunc
+        self.assertEqual(foo(a, b),
+                         a_(2) + a_(a_(2)) + b_(a_(2) + a_(a_(2))))
```
I think it would be good to move these to a separate `TestMiscIssues` test class to differentiate them from tests defined by design. I would imagine there may be more user issues in the future, quite a few features have a "user issues" test class like this already.

pearu replied:
I agree. Fixed.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+        for decor in all_func_kinds:
+            if decor in [pure_func]:
+                # Skip not-yet-supported functions
+                continue
```
If it's not supported, remove it!? (same comment throughout all these tests).

pearu replied:
Fixed.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+
+        def foo(f):
+            r = f(123)
+            return r
```
`return f(123)`

pearu replied:
Fixed.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+                b_ = decor(b)
+                self.assertEqual(njit(foo)(a_, b_), foo(a, b))
+
+    def test_in_ns_seq_call(self):
```
I feel that this test name needs expanding, or a doc string for what it is testing.

pearu replied:
Added doc string.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+                b_ = decor(b)
+                self.assertEqual(njit(mkfoo(b_))(a_), mkfoo(b)(a))
+
+    def test_ns_call(self):
```
I feel that this test name needs expanding, or a doc string for what it is testing.

pearu replied:
Added doc string.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+                a_ = decor(a)
+                self.assertEqual(njit(mkfoo(a_))(), mkfoo(a)())
+
+    def test_ns_out(self):
```
I feel that this test name needs expanding, or a doc string for what it is testing.

pearu replied:
Added doc string.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+                a_ = decor(a)
+                self.assertEqual(njit(mkfoo(a_))().pyfunc, mkfoo(a)())
+
+    def test_ns_call_out(self):
```
I feel that this test name needs expanding, or a doc string for what it is testing.

pearu replied:
Added doc string.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+        def mkfoo(a_):
+            def foo():
+                r = a_(123)
+                return r
```
`return a_(123)`

pearu replied:
Fixed.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+                self.assertEqual(njit(foo)((a_, b_), 2), foo((a, b), 2))
+
+
+class TestFuncionTypeExtensions(TestCase):
```
```suggestion
class TestFunctionTypeExtensions(TestCase):
```

pearu replied:
Fixed.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+                    libc = ctypes.cdll.msvcrt
+                elif os.name == 'posix':
+                    libpath = ctypes.util.find_library('c')
+                    print(f'libpath={libpath}')
```
remove print

pearu replied:
Fixed.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+
+class TestFuncionTypeExtensions(TestCase):
+
+    def _test_wrapper_address_protocol(self):
```
I feel that a variant of this test which talked to a `libm` function would be more convincing in that it can be asserted a known value is returned (for example, calling `libm::cos(0)`) opposed to relying on derived properties.

pearu replied:
I have added a new WAP test that calls `cos` and `sinf` from `libm`.

Enabling `time` test. However, IIRC, it failed in one of the CI nodes.

stuartarchibald commented on numba/types/function.py:
```diff
@@ -0,0 +1,116 @@
+            name = ftype.name
+        else:
+            raise TypeError(
+                f'constructing {type(self).__name__} from {type(ftype)}'
```
```suggestion
                f'constructing {type(self).__name__} from a {type(ftype)}'
```

pearu replied:
Fixed.

stuartarchibald commented on numba/types/function.py:
```diff
@@ -0,0 +1,116 @@
+                if isinstance(atype, Literal):
+                    atype = atype.literal_type
+                if not (atype <= ptype.atypes[i]):
+                    break
```
wondering if Optional needs handling?

pearu replied:
yes, likely, and there are other Numba types that could be supported here.
The currently supported Numba types are scalars, literals, arrays.

pearu replied:
It turns out that supporting `Optional` requires patching numba elsewhere:
```
numba.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of OptionalType(float64(float64)) i.e. the type 'float64(float64) or None' with parameters (float64)
No type info available for OptionalType(float64(float64)) as a callable.
[1] During: resolving callee type: OptionalType(float64(float64)) i.e. the type 'float64(float64) or None'
```
that cannot be fixed by handling Optional here. I suggest to leave it as TODO.

stuartarchibald commented on numba/types/function.py:
```diff
@@ -0,0 +1,116 @@
+            f'{self} argument types do not match with {args}')
+
+
+class FunctionProtoType(Type):
```
Perhaps `FunctionPrototype` as prototype is not two words?

pearu replied:
Fixed.

stuartarchibald commented on numba/types/function.py:
```diff
@@ -0,0 +1,116 @@
+class WrapperAddressProtocol(object):
+    """Base class for Wrapper Address Protocol.
+
+    Objects that type is derived from WrapperAddressProtocol can be
```
```suggestion
    Objects that inherit from the WrapperAddressProtocol can be
```

pearu replied:
Fixed.

stuartarchibald commented on numba/types/function.py:
```diff
@@ -0,0 +1,116 @@
+    """Base class for Wrapper Address Protocol.
+
+    Objects that type is derived from WrapperAddressProtocol can be
+    passed as arguments to numba njitted functions where it can be
```
```suggestion
    passed as arguments to Numba jit compiled functions where they can be
```
(assuming standard `jit` with object mode fall back works with this?

pearu replied:
Standard jit is not tested. Since the current PR supports fully only `cfunc` decorated functions as first-class types, the standard jit would not work for these anyway. So, I suggest leaving this to the next PR if that is ok with you?

Changing "numba njitted" to "Numba njit compiled".

pearu replied:
Tested object mode. Works out of the box after making CFunc callable.

stuartarchibald commented on numba/types/function.py:
```diff
@@ -0,0 +1,116 @@
+
+    Objects that type is derived from WrapperAddressProtocol can be
+    passed as arguments to numba njitted functions where it can be
+    used as first-class functions.  As minimum, the derived types must
```
```suggestion
    used as first-class functions.  As a minimum, the derived types must
```

pearu replied:
Fixed.

stuartarchibald commented on numba/types/function.py:
```diff
@@ -0,0 +1,116 @@
+        return self.name
+
+
+class WrapperAddressProtocol(object):
```
Should this class be abstract?

pearu replied:
Yes, now it inherits from ABC.

stuartarchibald commented on numba/types/function.py:
```diff
@@ -0,0 +1,116 @@
+            f'{type(self).__name__}.__wrapper_address__(sig) method')
+
+    def signature(self):
+        """Return a numba function type as Signature instance.
```
```suggestion
        """Return a Numba function type as a Signature instance.
```
Also, what sort of `Signature` instance? I think that this ought to be described.

pearu replied:
Rewrote the doc string as follows:
```
        Return the signature of a first-class function.

        Returns
        -------
        sig : Signature
          The returned Signature instance represents the type of a
          first-class function that the given WrapperAddressProtocol
          instance represents.
```

stuartarchibald commented on numba/typing/templates.py:
```diff
@@ -113,6 +113,12 @@ def as_function(self):
 
+    def as_type(self):
+        """
+        Convert signature to function type.
```
```suggestion
        Convert this signature to a function type.
```

pearu replied:
Fixed.

stuartarchibald commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+"""Provides numba type FunctionType that makes functions as instances
```
```suggestion
"""Provides a Numba type, FunctionType, that makes functions as instances
```

pearu replied:
Fixed.

stuartarchibald commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+"""Provides numba type FunctionType that makes functions as instances
+of a first-class function types.
```
```suggestion
of a first-class function type.
```

pearu replied:
Fixed.

stuartarchibald commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+import types as pytypes
```
This file could do with a more descriptive name, perhaps `test_first_class_functions.py` or `test_function_type.py`, it would also benefit from a docstring explaining what it is testing.

pearu replied:
Moved to `test_function_type.py`.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,209 @@
+of a first-class function types.
+"""
+# Author: Pearu Peterson
+# Created: December 2019
```
It's a habit of mine :) Will remove.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+
+def _get_wrapper_address(func, sig):
+    """Return the address of a compiled function that implements `func`
+    algoritm.
```
Fixed.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+
+
+def _get_wrapper_address(func, sig):
+    """Return the address of a compiled function that implements `func`
```
Fixed.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+
+    Warning: The compiled function must be compatible with the given
+    signature `sig`. If it is not, then calling the compiled function
+    will likely crash the program with segfault.
```
Rephased to "..then result of calling the compiled function is undefined."

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+    ----------
+    func : object
+      Specify a function object that can be numba.cfunc decorated or
+      an object that implements wrapper address protocol (see note below).
```
Fixed.

stuartarchibald commented:
Before this branch is merged, this feature needs to be well documented and some examples need writing as users are sure to ask. Thanks.

sklam commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+import types as pytypes
+from numba import njit, function, cfunc, types, int64, int32
```
The features in this PR depends on importing `function`. It should be auto-initialized. It can be done like: https://github.com/numba/numba/blob/ef119bcd1733ff49d71bdf2da8a66e91bb704f83/numba/targets/cpu.py#L65

pearu replied:
Fixed.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+      Specify a function object that can be numba.cfunc decorated or
+      an object that implements wrapper address protocol (see note below).
+    sig : Signature
+      Specify function signature.
```
Fixed.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+    Parameters
+    ----------
+    func : object
+      Specify a function object that can be numba.cfunc decorated or
```
The `func` can also be Python function. When this is the case, the mapping `func = cfunc(sig)(func)` will be applied.

`func` must be compatible with `sig`, in general, see the Warning note. There are no compatibility checks in order to keep `_get_wrapper_address` light-weight. In fact, compatibility is ensured at the point of calling the njitted function with function arguments so that `_get_wrapper_address` is good (in most cases).

pearu replied:
Added more notes to documentation string.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+    Returns
+    -------
+    addr : int
+      Address (pointer value) of a compiled function.
```
Fixed.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+    ------------------------------
+
+    An object implements the wrapper address protocol iff the object
+    provides a callable attribute named __wrapper_address__ that takes
```
Fixed.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+            f'wrapper address must be integer, got {type(addr)} instance')
+    if addr <= 0:
+        raise ValueError(f'wrapper address of {type(func)} instance must be'
+                         f' positive integer but got {addr}')
```
Fixed.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+def unbox_function_type(typ, obj, c):
+    sfunc = cgutils.create_struct_proxy(typ)(c.context, c.builder)
+
+    # Get obj wrapper address. The code below trusts that the function
```
Fixed.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+    # Get obj wrapper address. The code below trusts that the function
+    # numba.function._get_wrapper_address exists and can be called
+    # with two arguments. However, if an exception is raised in the
+    # function, then it will be catched and propagated to the caller.
```
Fixed.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+        addr = pyval._wrapper_address
+        sfunc = cgutils.create_struct_proxy(typ)(context, builder)
+        llty = context.get_value_type(nbtypes.voidptr)
+        sfunc.addr = builder.inttoptr(ir.Constant(ir.IntType(64), addr), llty)
```
In a 32-bit system, an `addr` is a 32-bit memory address (of type Python `int`).
Apparently, `inttoptr` can handle casting 64-bit integer to `voidptr` as the value never larger than `2**32-1`.

But I agree, the code should use `IntType(32)` on 32-bit systems.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+            # decref should be called?
+            # TODO: see
+            # https://github.com/numba/numba/blob/77cb53ba8966a38bfbd3413559e4f04b12812535/numba/targets/boxing.py#L505
+            # as an alternative way for boxing
```
Re TODO 1: this is obsolete, will remove.
Re TODO 2: this requires a test case that makes the current code fail. Whether such a test case exists, is not known atm.
Re TODO 3: the cited reference is not applicable (it would mean pickling Python functions but we need the address of the original function, not of the copy). Will remove.

I'll keep this conversation open until having more info re TODO 2.

pearu replied:
tested TODO 2 -> not an issue (pyaddr is always valid as python functions are attached to CFun instances).

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+            c.pyapi.err_set_string("PyExc_MemoryError", cstr)
+            c.builder.ret(c.pyapi.get_null_object())
+        with orelse:
+            pass
```
Using now `if_then` as `else` is not needed due to `return`. Fixed.

pearu commented on numba/pythonapi.py:
```diff
@@ -475,6 +475,11 @@ def long_as_longlong(self, numobj):
+    def long_as_long(self, numobj):
+        fnty = Type.function(self.ulonglong, [self.pyobj])
+        fn = self._get_function(fnty, name="PyLong_AsLong")
+        return self.builder.call(fn, [numobj])
```
`long_as_long` is not used anymore but it has been tested in a working code.
I will remove it unless you want to keep it. Advice?

pearu replied:
Removed.

pearu commented on numba/pythonapi.py:
```diff
@@ -568,6 +573,11 @@ def long_from_unsigned_int(self, ival):
+    def long_from_voidptr(self, ival):
+        fnty = Type.function(self.pyobj, [self.voidptr])
+        fn = self._get_function(fnty, name="PyLong_FromVoidPtr")
+        return self.builder.call(fn, [ival])
```
`long_from_voidptr` is not used anymore but it has been tested in a working code.
I will remove it unless you want to keep it. Advice?

pearu replied:
Removed.

pearu commented on numba/function.py:
```diff
@@ -0,0 +1,204 @@
+    provides a callable attribute named __wrapper_address__ that takes
+    Signature instance as the argument, and returns an integer
+    representing the address or pointer value of a compiled function
+    with given signature.
```
Fixed.

pearu commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+    print('{"":*^70}')
+
+
+# Decorators for tranforming a Python function to different kinds of functions:
```
Fixed.

pearu commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+    return f
+
+
+def njit2_func(func, sig=int64(int64)):
```
`njit_with_sig_func` is more accurate. Renaming.

sklam commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+
+# Decorators for tranforming a Python function to different kinds of functions:
+def pure_func(func, sig=int64(int64)):
+    assert isinstance(func, pytypes.FunctionType), repr(func)
```
This code is not exercised. I would prefer not to include code that is never used.

pearu replied:
Removed.

sklam commented:
With this PR, i think `CFunc` needs to be callable, otherwise we break the contract that code still works without the decorator. 



<details>
<summary>Sample code that fails without `@njit`</summary>
```python
import numpy as np
from numba import njit, cfunc, intp, float64, function

@cfunc("intp(intp, float64)")
def foo1(x, y):
    return x + y


@cfunc("intp(intp, float64)")
def foo2(x, y):
    return x - y

# @njit   # doesn't work without @njit
def bar(fx, fy, i):
    a = np.array([10], dtype=np.intp)
    b = np.array([12], dtype=np.float64)
    if i == 0:
        f = fx
    elif i == 1:
        f = fy
    else:
        return
    return f(a[0], b[0])    # TypeError: 'CFunc' object is not callable

r = bar(foo1, foo2, 0)

print(r)
```

</details>

sklam commented on numba/types/function.py:
```diff
@@ -0,0 +1,116 @@
+            for i, atype in enumerate(args):
+                if isinstance(atype, Literal):
+                    atype = atype.literal_type
+                if not (atype <= ptype.atypes[i]):
```
Not all `Type` instances are ordered. 

Below is a failing case because `Array` is not ordered:

<details>
<summary>example code</summary>

```python
# Failed because Array type-class doesn't have <= comparison
import numpy as np
from numba import njit, cfunc, intp, float64, function

@cfunc("intp(intp[:], float64[:])")
def foo1(x, y):
    return x[0] + y[0]


@cfunc("intp(intp[:], float64[:])")
def foo2(x, y):
    return x[0] - y[0]

@njit
def bar(fx, fy, i):
    a = np.array([10], dtype=np.intp)
    b = np.array([12], dtype=np.float64)
    if i == 0:
        f = fx
    elif i == 1:
        f = fy
    else:
        return
    return f(a, b)

r = bar(foo1, foo2, 0)
print(r)
```
</details>

pearu replied:
Fixed. The currently supported argument types are arrays, numbers, and literals.

pearu commented on numba/tests/test_function.py:
```diff
@@ -0,0 +1,575 @@
+
+class TestFuncionType(TestCase):
+
+    def _test_issue_3405_original(self):
```
Removed the test for now.

pearu commented:
@stuartarchibald and @sklam  thank you for your exhaustive review on this PR.

I have addressed your feedback as follows:
1. Added "Functions" subsection to Numba reference manual with examples
2. Added `Array` support
3. Made `CFunc` callable
4. For testing WAP, demonstrate calling `cos` and `sinf` from the standard math library
5. Revised unit tests, added docstrings
6. Ensured that both `njit` and `jit(forceobj=True)` compiled functions are supported
7. Fixed many typos and language errors

I'll continue addressing two TODO items (incref question and Optional) but the rest of the PR is ready for review.

esc commented:
@pearu There may be some issues with using single backticks in docs for marking code. IIRC for the Numba sphinx setup you need to use double backticks instead. I recommend running sphinx locally and checking the respective output.

stuartarchibald commented:
Thanks for the fixes so far, I've done some light testing here: https://github.com/stuartarchibald/numba/commit/06c08835ce6988d9ca477f4db97053805eb3c9ba there are a number of issues that I feel need discussion/further work and I'm of the view it'd be good to spend some time testing it further and fixing usability things/generally stressing the implementation.

Further, @sklam and I caught up over this a few minutes ago when discussing today's 0.48.0RC work plan, we think that:
1. We'd like to spend some more time on this to add more extensive testing, especially tests that specifically check for things added/made possible by this feature.
2. We need to check this feature with respect to caching to ensure that there's no way addresses are ending up baked in to the machine code (a problem we've noticed previously when doing similar things).
3. We'd like to review the documentation and the code again in light of the fixes/changes pushed recently.

As a result of 1-3 and that we have to start the mechanism for releasing 0.48.0 now, I'm sorry to have to disappoint you, but I'm afraid it's not going to make it into the release, however, we will endeavour to get this into 0.49.0dev0 very early on. Users wanting to try this out once merged will of course be able to use 0.490dev0 builds from the Numba channel. Thanks for your understanding.

pearu commented:
@stuartarchibald thanks for testing and stressing this PR.
I have added some comments to your test cases, here's a summary:
test_01 - the failure cannot be avoided because of the nature of `cfunc` compilation.
test_02, test_03, test_04, test_06, test_11 - the issues are out of the scope of this PR 
test_05, test_07, test_08, test_10 - these tests work now
test_09 - the support for Optional types needs to be implemented.

Re point 2: the addresses are not baked into machine code when first-class functions are passed in as arguments to Numba jit compiled functions. A simple check of it would be to disable the compilation of the jit compiled function and test it with different function arguments. As a simple proof, the addresses are resolved in runtime by the call to `_get_wrapper_address` when unboxing the first-class function arguments.

When the first-class functions are passed in via namespace scoping, the corresponding functions are considered as constants and are resolved at the stage of `cfunc` compilation. If this is not a desired behavior then some sort of callback to Python process will be required in order to resolve the scope of the first-class functions.

sklam commented on numba/function.py:
```diff
@@ -0,0 +1,198 @@
+        addr = pyval._wrapper_address
+        sfunc = cgutils.create_struct_proxy(typ)(context, builder)
+        llty = context.get_value_type(nbtypes.voidptr)
+        sfunc.addr = builder.inttoptr(ir.Constant(
```
Can you switch the use of `builder.inttoptr` to `context.add_dynamic_addr`? Doing so will stop the compilation unit being cached to disk.

sklam replied:
This is only needed for freezing Constant pointers that are only valid in the current process.

pearu replied:
Done.

sklam commented on numba/function.py:
```diff
@@ -0,0 +1,198 @@
+        llty = context.get_value_type(nbtypes.voidptr)
+        sfunc.addr = builder.inttoptr(ir.Constant(
+            ir.IntType(context.address_size), addr), llty)
+        sfunc.pyaddr = builder.inttoptr(
```
here as well

pearu replied:
Done.

sklam commented on numba/function.py:
```diff
@@ -0,0 +1,198 @@
+    # TODO: reconsider the boxing model, see the limitations in the
+    # unbox function above.
+    pyaddr_ptr = cgutils.alloca_once(c.builder, c.pyapi.pyobj)
+    raw_ptr = c.builder.inttoptr(sfunc.pyaddr, c.pyapi.pyobj)
```
`builder.inttoptr` -> `context.add_dynamic_addr`

sklam commented on numba/core/types/function.py:
```diff
@@ -0,0 +1,130 @@
+    """
+
+    @abstractmethod
+    def __wrapper_address__(self, sig):
```
Why does this method need to take a `Signature` argument? In what case would `sig != self.signature`?

pearu replied:
The WAP object may hold addresses to different functions. See, for example
https://github.com/Quansight/numba/blob/pearu/3405/numba/tests/test_function_type.py#L532-L559

That said, I can see that `sig` argument could be eliminated for `CFunc` objects. Also in the case of WAP objects, the `sig` argument is used only for ensuring signature compatibility that holds always except in possible bugs in user-code. In fact, the example referenced above could be redesigned such that `__wrapper_address__` would not need the `sig` argument.

However, the situation for eliminating `sig` argument is not clear in the case of `Dispatcher` objects (`njit` decorated functions) where the `__wrapper_address__` may need to compile the `sig`-specific function in runtime.

I think this question requires research with respect to supporting `njit` decorated functions that I am currently working on. To enable both ways (until the questions is resolved), we could use
```
@abstractmethod
def __wrapper_address__(self, sig=None):
```
for defining the interface. What do you think?

pearu replied:
I can now confirm that `sig` is not required for njit-decorated first-class functions.

pearu replied:
The signature argument is now removed from `__wrapper_address__`.

pearu commented:
Current (local) status of njit-decorated first-class function support:
```
    @nb.njit
    def a(x):
        return x + 1

    @nb.njit
    def b(x):
        if x < 0:
            raise ValueError('expected non-negative input')
        return x + 2

    @nb.njit
    def foo(f, x):
        return f(x)

    assert foo(a, 0) == 1  # compiles foo and a
    assert foo(b, 0) == 2  # compiles b
    assert foo(a, 1) == 2  # no compilations
    assert foo(b, 1) == 3  # no compilations

    r = foo(b, -5)  # prints exception to stderr, r value is arbitrary (0)

    ##################
    @nb.njit
    def c(x):
        return x + 3

    @nb.njit
    def bar(f, x):
        return x

    print(bar(a, 123))  # compiles bar, reuses `a` compile result
    print(bar(c, 567))  # compiles c
    # calling bar(c, 567) before (bar(a, 123) results failure because there is no type info re `c`. Ideally, `c` should be ignored..
```

pearu commented:
@sklam @stuartarchibald, please review:
- rebased to master,
- removed signature from `__wrapper_address__` method.

pearu commented:
Note that PR #5287 adds the jit-decorated functions support on the top of this PR.
So, if preferred, this PR can be shelved and we can proceed working with PR #5287 only.

sklam commented:
I have some suggestions at https://github.com/numba/numba/pull/5307.  https://github.com/numba/numba/pull/5307/commits/8156a197554e8612534467b27e03110db8520565 would simplify the logic for `.get_call_type()`'s casting rule.

pearu commented:
> I have some suggestions at #5307. [8156a19](https://github.com/numba/numba/commit/8156a197554e8612534467b27e03110db8520565) would simplify the logic for `.get_call_type()`'s casting rule.

Thanks, applied the suggestion to #5287 .

stuartarchibald commented:
Note to self. Check docs. @sklam approves code.

stuartarchibald commented:
If https://github.com/numba/numba/pull/5287 is replacing, can this be closed now?

pearu commented:
yes, this can be closed.
